src/Hl7.Fhir.Core/Serialization/SerializationUtil.cs

```diff
@@ -0,0 +1,485 @@
+            {
+                XmlWriterSettings settings = new XmlWriterSettings
+                {
+                    Encoding = new UTF8Encoding(false),
+                    OmitXmlDeclaration = true,
+                    NewLineHandling = NewLineHandling.Entitize
+                };
+
+                using (XmlWriter xw = XmlWriter.Create(stream, settings))
+                {
+                    serializer(xw);
+                    xw.Flush();
+                    return stream.ToArray();
+                }
+            }
+        }
+
+        public static string WriteXmlToString(Action<XmlWriter> serializer, bool pretty = false)
+        {
+            StringBuilder sb = new StringBuilder();
+
+            XmlWriterSettings settings = new XmlWriterSettings
+            {
+                OmitXmlDeclaration = true,
+                NewLineHandling = NewLineHandling.Entitize,
+                Indent = pretty
+            };
+
+            // [WMR 20160421] Explicit disposal
+            using (XmlWriter xw = XmlWriter.Create(sb, settings))
+            {
+                serializer(xw);
+                xw.Flush();
+                return sb.ToString();
+            }
+        }
+
+        public static XDocument WriteXmlToDocument(Action<XmlWriter> serializer)
+        {
+            var doc = new XDocument();
+
+            using (XmlWriter xw = doc.CreateWriter())
+            {
+                serializer(xw);
+                xw.Flush();
+            }
+
+            return doc;
+        }
+
+        /// <summary>
+        /// Replace all the XML specific special characters with the XHTML equivalents
+        /// </summary>
+        /// <remarks>
+        /// this is based on own research plus combining with results shown here:
+        /// http://www.codeproject.com/Articles/298519/Fast-Token-Replacement-in-Csharp
+        /// The RegEx approach does not require multiple passes or string creations
+        /// while replacing all the items.
+        /// It occurs in O(n) StringBuilder concatenations + O(n) dictionary lookups
+        /// </remarks>
+        /// <param name="xml"></param>
+        /// <returns></returns>
+        public static string SanitizeXml(string xml)
+        {
+            if (string.IsNullOrEmpty(xml))
+                return xml;
+
+            Dictionary<string, string> xr = getXmlReplacements();
+
+            //s.Reset();
+            //s.Start();
+            string resultRE;
+            var matches = _re.Matches(xml);
+            if (matches.Count > 0)
+            {
+                StringBuilder sbRE = new StringBuilder();
+                int currentPosition = 0;
+                foreach (Match m in matches)
+                {
+                    if (xr.ContainsKey(m.Value))
+                    {
+                        sbRE.Append(xml.Substring(currentPosition, m.Index - currentPosition));
+                        sbRE.Append(xr[m.Value]);
```

Comment:
We should do `xr.TryGetValue` instead of blindly looking it up with the indexer. 

Then we could handle bad/unknown escape sequences (e.g., `&foobar;`) better than just throwing a `KeyNotFoundException` which isn't helpful for the client to figure out why their XML is bad.